tests/cli/test_compose_alibaba.sh

```diff
@@ -0,0 +1,197 @@
+                        --DiskDeviceMapping.1.Format qcow2 \
+                        --DiskDeviceMapping.1.OSSBucket $ALICLOUD_BUCKET \
+                        --DiskDeviceMapping.1.OSSObject $IMAGE \
+                        --ImageName $IMAGE"
+
+        # wait for status to become available
+        while [ `$ALI_DIR/aliyun ecs DescribeImages --ImageName $IMAGE --Status Available | jq .Images.Image | jq -r '.[0].ImageName'` == "null" ]; do
+            rlLogInfo "Waiting for import to complete ..."
+            sleep 60
+        done
+
+        rlRun -t -c "$ALI_DIR/aliyun ecs DescribeImages --ImageName $IMAGE"
+        IMAGE_ID=`$ALI_DIR/aliyun ecs DescribeImages --ImageName $IMAGE | jq .Images.Image | jq -r '.[0].ImageId'`
+
+        if [ "$IMAGE_ID" == "null" ]; then
+            rlFail "IMAGE_ID is empty!"
+        else
+            rlLogInfo "IMAGE_ID=$IMAGE_ID"
+        fi
+    rlPhaseEnd
+
+    rlPhaseStartTest "Start ECS instance"
+        INSTANCE_TYPE="ecs.n1.medium"
+
+        # generate & import new ssh key
+        KEY_NAME=Composer-Test-$UUID
+        SSH_KEY_DIR=`mktemp -d /tmp/composer-ssh-keys.XXXXXX`
+        rlRun -t -c "ssh-keygen -t rsa -N '' -f $SSH_KEY_DIR/id_rsa"
+        SSH_PUB_KEY=$(cat $SSH_KEY_DIR/id_rsa.pub)
+        rlRun -t -c "$ALI_DIR/aliyun ecs ImportKeyPair --KeyPairName $KEY_NAME --PublicKeyBody '$SSH_PUB_KEY'"
+
+        RELEASE_TIME=$(date -u -d "24 hours" '+%FT%TZ')
+
+        # SecurityGroup is composer-allow-ssh
+        # VPC is composer-vpc
+        response=$($ALI_DIR/aliyun ecs RunInstances --Amount 1 --ImageId $IMAGE_ID \
+                    --InstanceType=$INSTANCE_TYPE --InstanceName Composer-Test-VM-$UUID \
+                    --SecurityGroupId sg-0xi4w9isg0p1ytj1qbhf \
+                    --VSwitchId vsw-0xi36w0a9l894vf2momfb \
+                    --KeyPairName $KEY_NAME \
+                    --InternetMaxBandwidthIn 5 --InternetMaxBandwidthOut 5 \
+                    --AutoReleaseTime $RELEASE_TIME)
+        rlAssert0 "VM started successfully" $?
+        rlLogInfo "$response"
+
+        INSTANCE_ID=`echo "$response" | jq .InstanceIdSets.InstanceIdSet | jq -r '.[0]' `
+
+        until [ $($ALI_DIR/aliyun ecs DescribeInstanceAttribute --InstanceId $INSTANCE_ID | jq -r .Status | grep "Running\|Stopped") ]; do
+            sleep 30
+            rlLogInfo "Waiting for instance to start ..."
+        done
+
+        rlAssertEquals "Instance $INSTANCE_ID is Running" \
+            "$($ALI_DIR/aliyun ecs DescribeInstanceAttribute --InstanceId $INSTANCE_ID | jq -r .Status)" "Running"
+        rlRun -t -c "$ALI_DIR/aliyun ecs DescribeInstanceAttribute --InstanceId $INSTANCE_ID"
+
+        IP_ADDRESS="null"
+        while [ "$IP_ADDRESS" == "null" ]; do
+            rlLogInfo "IP_ADDRESS is not assigned yet ..."
+            sleep 30
+            IP_ADDRESS=`$ALI_DIR/aliyun ecs DescribeInstanceAttribute --InstanceId $INSTANCE_ID | jq -r .PublicIpAddress.IpAddress | jq -r '.[0]'`
+        done
```

Comment:
I can see the `IP_ADDRESS` already assigned when line 161 is being run. Maybe it can be retrieved from there and check if it's `null` instead of setting it to `null`,  waiting and then retrieving it?

Reply:
line 161 is like an info dump after we're certain the VM instance is in status running so we've got a reference to all of its attributes.

Then the script proceeds to get the IP address. I can instead assign the output of line 161 to a variable, rlLogInfo that variable and then echo & grep for the IP address but it is not any different.  So I'm not inclided to do that unless you insist too much.

Reply:
Ok good, I understand line 161, but what I did not understand is why not do line 163 the same way like line 167, why assing it `null` first? But yes the results would be the same, so you don't have to change it I was just curious.